### ISSUE TITLE
[FR] Change scope of private variables to protected in UrlRule

### ISSUE THREAD

**Issue by fabioaccetta** [yiisoft/yii2#7204]
I notice that the UrlRule class has some private variables, but some of them are used inside init() method; this way, extending UrlRule become unuseful.

For example, I'm developing a custom url rule that needs only small changes to UrlRule methods - I need to override only createUrl() and parteRequest(), init() is what I need - but I can't simply extend UrlRule because I need to implement also init() because of its private variables.

In other words, I think that extending UrlRule makes sense only if the child class can access the variables initialized inside init(), so do you agree in changing the scope of these variables?

**Comment by cebe** [comment]
please show how exactly you need to extend it and which variables you need. As far as I see it may be useful to split init() into multiple functions that could then be overridden instead of making the properties protected.

**Comment by fabioaccetta** [comment]
The problem was: my own rule must support patterns that contains also slash characters (with `'enablePrettyUrl' => 'true'`).

In `yii1` i was forced to implement my own rule (with `createUrl()` and `parseRequest()`) because generated urls were encoded (then with `/` converted to `%2F`).

While porting my project to `yii2` i have faced the same problem; so, as done for `yii1`, I was trying to use the variable `UrlRule::$_template` (well initialized during `init()`).

However i have discovered that it's no more needed because today I found the new property `UrlRule::$encodeParams` that makes me able to use `UrlRule` as is, without the need to override it, solving my problem.

As a result, i have no more examples to support the need to make private variables protected, so i suppose this feature request becomes meaningless, sorry.
